### ISSUE TITLE
Make NixOS module to update apropos database

### ISSUE THREAD

**Issue by domenkozar** [NixOS/nixpkgs#14472]
Using `makewhatis` command.

**Comment by jerith666** [comment]
What's the difference between `makewhatis` and `mandb`?  Starting with `man man`, I am lead to `apropos` and then `mandb`, not to `makewhatis` ... ?

**Comment by jerith666** [comment]
As an aside, `man -K` is a fair substitute for `apropos` / `man -k`, and since it does a brute-force search, it works even in the absence of the index / DB under discussion here.  On my SSD-based system it's actually pretty speedy, too.

**Comment by domenkozar** [comment]
> whatis searches a set of database files containing short descriptions of system commands for keywords and displays the result on the standard output. 

> Often a wrapper for the man -k command, the apropos command is used to search all manual pages for the string specified. This is often useful if one knows the action that is desired, but does not remember the exact command or page name.

> makewhatis extracts the name and a short description from unformatted manpages and creates the whatis(1) database

**Comment by jerith666** [comment]
On my `nixos-unstable` system, both `man whatis

**Comment by jerith666** [comment]
On my `nixos-unstable` system, both `man whatis` and `man apropos` say, respectively:

> `index`  databases  are  used during the search, and are updated by the `mandb` program.

> The database searched by `apropos` is updated by the `mandb` program. 

I just don't want us to spend effort implementing support for a deprecated manpage indexer.  :)

**Comment by domenkozar** [comment]
@jerith666 I don't really understand what you're trying to say. `apropos` offers search for an index, I don't see any equivalent in `man`.

**Comment by jerith666** [comment]
`makewhatis` and `mandb` seem to be two alternate means of doing the same thing: generating an index that enables `apropos` and `whatis` searches.  The main point I'm trying to make is that `mandb` seems to be the newer, preferred way of building the index.  `makewhatis`, the program you suggested in your initial description of this issue, seems to be deprecated -- it looks like the current versions of `apropos` and `whatis` would probably work with the index that it produces, but future versions may not.

Since my initial attempts to run either `makewhatis` or `mandb` on nixos were unsuccessful (both seem to only want to write into the manpage hierarchy, neither provides an option to write to an arbitrary destination directory), it looks like it's going to be a non-trivial task to get one of them working.  So I want to make sure we spend that effort on the current, supported one.  :smile: 

However, I should also say that I knew almost nothing about how the suite of `man`-related programs actually worked until about a week ago, and the above is just what I've gathered from poking around on http://man-db.nongnu.org/ etc. -- so it's entirely possible that I'm missing something obvious ...

(On my old gentoo system, FWIW, there was a daily cron job to run `makewhatis`.)

**Comment by domenkozar** [comment]
@jerith666 so we should rather use `mandb` and `man`. Makes sense, thanks!

**Comment by RamKromberg** [comment]
@domenkozar I fixed up the search paths in https://github.com/NixOS/nixpkgs/issues/18521 so the system and user database build correctly now. Note the instructions for the user db especially.

Also, Ideally, I think _mandb_ and _mandb -u_ should be executed following package installations (_nixos-rebuild_ and _nix-env_).

**Comment by jerith666** [comment]
I figured they would be executed *only* on package installations.  Unlike a regular distro, we know those are the only times that new man pages can be introduced.  We should be able to put the output in the nix store rather than under `/var`.  Though as far as I can tell this is likely to require patching `mandb`.  :-(

**Comment by RamKromberg** [comment]
@jerith666 Putting the output in a store and writing a proper module is pretty straightforwards:
1. In postInstall, delete _man_db.conf_ and place a symbolic link to _/etc/man_db.conf_.
2. Write a system module that modifies _/etc/man_db_'s MANDB_MAP to point to an output store and executes mandb.
3. Write a user module that modifies _~/.manpath_ similarly to _/etc/man_db_ (see my PR).
4. Write a user module that executes _mandb -u_.

**Comment by RamKromberg** [comment]
Btw, I've messed around with _mandoc_ but I'm not 100% sold out on it. I suspect once upstream releases a version without _sqlite_ as a dependency, someone, possibly myself, will port it to _Nix_ and try replace _man-db_ + _groff_ + _db_ for _mandoc_ on closure size and compile times rational alone much like _Alpine_ and _Void_ already did. You can play around with it using this (no sqlite \ apropos functionality. just to test the rendering):
```
{ stdenv, fetchurl, zlib, utillinux }:

stdenv.mkDerivation rec {
  name = "mandoc-${version}";
  version = "1.13.4";

  src = fetchurl {
    url = "http://mdocml.bsd.lv/snapshots/mdocml-${version}.tar.gz";
    sha256 = "1vz0g5nvjbz1ckrg9cn6ivlnb13bcl1r6nc4yzb7300qvfnw2m8a";
  };

  buildInputs = [ zlib utillinux ];

  configureLocal = ''
    HAVE_WCHAR=1
    MANPATH_DEFAULT="/run/current-system/sw/share/man"
    PREFIX="$out"
    HAVE_MANPATH=1
    MANM_MANCONF="mandoc.conf"
    MANM_MAN="mandoc_man"
    MANM_MDOC="mandoc_mdoc"
    MANM_ROFF="mandoc_roff"
    MANM_EQN="mandoc_eqn"
    MANM_TBL="mandoc_tbl"
    BINM_MAN=mman
    BINM_SOELIM=msoelim
    LD_OHASH="-lutil"
    BUILD_DB=0
    BINM_APROPOS=mapropos
    BINM_WHATIS=mwhatis
    BINM_MAKEWHATIS=mandocdb
  '';

  preConfigure = ''
    echo $configureLocal > configure.local
  '';

  meta = with stdenv.lib; {
    homepage = "http://mdocml.bsd.lv/";
    description = "suite of tools compiling mdoc and man.";
    license = licenses.bsd3;
    platforms = platforms.linux;
  };
}
```

While the man pages turn out fine, the pdf and html outputs erroneous with my test table.

Regardless, keep this all mind if you're planning to do a module. ;)

**Comment by RamKromberg** [comment]
Since my mandoc efforts aren't going anywhere, I'll probably be sticking to stable for a while so a few things to note for anyone interested in doing a module:
1. man-db has a setuid option that should be coupled with a man user and service for building the system database.
2. the database is built once for the system (or the man user\group), and once for the user's packages. The building is done differently for each (_mandb_ _mandb -u_) to eliminate duplicating the system entries.

Overall, just avoid running man-db \ groff as root in a module and it should safe enough to build docs. The user data-base is still open to a sneaky ```.sy``` instructions coupled with a groff deference that would allow triggering the unsafe instructions (it's a big non web facing package so there's a couple of those)... But it can't be helped.

Good luck :)

**Comment by devrandomguy** [comment]
I don't suppose someone could point me to a package, that would serve as a good example of how manpages should currently be distributed by all Nix packages?

Right now, I am figuring out how to get my host system's man-db to index the tmux manpage, after a `nix-env -i tmux`. `mandb` searches for pages in `$HOME/.nix-profile/share/man`, presumably it discovers this based on the `$HOME/.nix-profile/bin` in `$PATH`. However, it seems that the only packages that place any manpages here, are the actual Nix tools themselves. Should all packages symlink their manpages from `$HOME/.nix-profile/share/man`, during install, and then call `mandb`?

Or, as a user, should I install Nix's `man-db` package, and use that instead of the host system's mandb?

**Comment by jerith666** [comment]
I'm not sure how `man` is supposed to work with `nix` on a non-NixOS system.  All I can say is that I have this on my NixOS system:

```
lrwxrwxrwx 1 root root 81 Dec 31  1969 /run/current-system/sw/share/man/man1/tmux.1.gz -> /nix/store/qix8p6ipq0nsjx967jzikq80bxmayj1p-tmux-2.3-man/share/man/man1/tmux.1.gz
```

Maybe 94627e658beabf2b4c766d31afa273a55879b46a is relevant?

**Comment by mmahut** [comment]
Are there any updates on this issue, please

**Comment by l0b0** [comment]
FYI, it looks like the command in [the wiki page](https://nixos.wiki/wiki/Apropos) is borken:

```text
$ sudo su -l root -c mandb
Processing manual pages under /run/current-system/sw/share/man...
/run/current-system/sw/bin/mandb: warning: cannot create catdir /var/cache/man/nixos
/run/current-system/sw/bin/mandb: can't create index cache /var/cache/man/nixos/6356: No such file or directory
/run/current-system/sw/bin/mandb: can't update index cache /var/cache/man/nixos/6356: No such file or directory
done.
```

**Comment by DrPyser** [comment]
```
# nixos-version
19.09.1594.fbe321e6669 (Loris)

# su -l root -c mandb
Processing manual pages under /run/current-system/sw/share/man...
/run/current-system/sw/bin/mandb: warning: cannot create catdir /var/cache/man/nixos
/run/current-system/sw/bin/mandb: can't create index cache /var/cache/man/nixos/29002: No such file or directory
/run/current-system/sw/bin/mandb: can't update index cache /var/cache/man/nixos/29002: No such file or directory
done.
```
Manually creating `/var/man/cache/nixos` fixes this.
From the [analogous issue for `locate`](https://github.com/NixOS/nixpkgs/issues/17397), it seems a systemd service is used to setup the locatedb. Should this be done for mandb as well?

**Comment by jerith666** [comment]
I think `locate` is a bit of a false analogy.  It indexes impure things outside the nix store, so it makes sense for it to have a service that updates its backing DB periodically.

`apropos` and `whatis`, conversely, are only about searching man pages for installed commands.  Since the set of installed commands is fixed under `nix` (only changed by `nixos-update` and `nix-env`), we should be able to create a fixed backing DB for each generation / profile (when each of those two commands is run) and never need a service to update it.

**Comment by JustinLovinger** [comment]
The current lack of `apropos` support breaks tab completion in the `fish` shell. Also `mandb` does not index user man pages.

**Comment by DamienCassou** [comment]
I think this issue got fixed with #86489.

**Comment by DamienCassou** [comment]
I think this issue got fixed with #86489.
